[extensions/observer/cfgardenobserver] Implement cfgardenobserver

tomasmota opened the pull request:
**Description:** 
First Component PR: #33727
This is the second PR for adding the cfgardenobserver, with the first suggested implementation. There are definitely some decisions made that require feedback, such as adding the CloudFoundry application labels to the Endpoint labels, and the decision to use the `Container` EndpointType at all.

**Link to tracking Issue:** #33618 

**Testing:** Unit testing of config and extension

**Documentation:** Updated readme with new configuration and endpoints

tomasmota committed:
[extensions/observer/cfgardenobserver] Implement cfgardenobserver

tomasmota committed:
[extensions/observer/cfgardenobserver] Updated docs

tomasmota committed:
[extensions/observer/cfgardenobserver] Added changelog

tomasmota committed:
[extensions/observer/cfgardenobserver] Tidy after dep update

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Make including app labels optional

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Run tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

crobert-1 commented:
```
Generated code is out of date, please run "make generate" and commit the changes in this PR.
```

tomasmota committed:
[extensions/observer/cfgardenobserver] Gen

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go Tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Add include_app_labels to docs

crobert-1 commented on extension/observer/cfgardenobserver/README.md:
```diff
@@ -20,34 +20,61 @@ The `cfgarden_observer` looks at the current host to discover Garden containers.
-    endpoint: my/path/to/garden.sock
     # determines how often to look for changes in endpoints.
     refresh_interval: 30s
+    # determines how often app metadata cache is refreshed
```
Just a nit, since we have the table below that explains each configuration option, we probably don't need a comment for each option in this config example.

crobert-1 commented on extension/observer/cfgardenobserver/config.go:
```diff
@@ -4,15 +4,116 @@
+	ClientSecret string `mapstructure:"client_secret"`
 }
+
+// AuthType describes the type of authentication to use for the CloudFoundry API
+type AuthType string
+
+const (
+	// AuthTypeClientCredentials uses a client ID and client secret to authenticate
+	AuthTypeClientCredentials AuthType = "client_credentials"
+	// AuthTypeUserPass uses username and password to authenticate
+	AuthTypeUserPass AuthType = "user_pass"
+	// AuthTypeToken uses access token and refresh token to authenticate
+	AuthTypeToken AuthType = "token"
```
We should make as many of these private as possible.

tomasmota replied:
I have made the consts private, should AuthType also be private?

crobert-1 replied:
Yes, we should make `AuthType` private as well, if possible.

crobert-1 commented on extension/observer/cfgardenobserver/config.go:
```diff
@@ -4,15 +4,116 @@
+	}
+
+	c := config.CloudFoundry
+	if c.Endpoint == "" {
+		return errors.New("config.Endpoint must be specified when include_app_labels is set to true")
+	}
+	if c.AuthType == "" {
+		return errors.New("config.AuthType must be specified when include_app_labels is set to true")
+	}
```
The error messages don't seem to align with the README table for config options. The table states that these options are simply required, without any conditional placed on `include_app_labels`. I may be missing something, but which one is correct?

tomasmota replied:
You're right, I added the `include_app_labels` option in the end after opening the PR, and forgot to adjust the README

crobert-1 commented on extension/observer/cfgardenobserver/config.go:
```diff
@@ -4,15 +4,116 @@
+			return fieldError(AuthTypeToken, "refresh_token")
+		}
+	default:
+		return fmt.Errorf("unknown auth_type: %s", c.AuthType)
```
```suggestion
		return fmt.Errorf("configuration option `auth_type` must be set to one of the following values: [user_pass, client_credentials, token]. Specified value: %s", c.AuthType)
```
It would be good to show that it's required, what values are valid, and what the user specified here.

crobert-1 commented on extension/observer/cfgardenobserver/README.md:
```diff
@@ -20,34 +20,61 @@ The `cfgarden_observer` looks at the current host to discover Garden containers.
-| endpoint         | string | unix:///var/vcap/data/garden/garden.sock | The endpoint to connect to the Garden API.             |
+| Name                        | Type   | Default                           | Description                                                       |
+| --------------------------- | ------ | --------------------------------- | ----------------------------------------------------------------- |
+| refresh_interval            | string | 60s                               | Determines how often to look for changes in endpoints.            |
```
```suggestion
| refresh_interval            | string | 1m                                | Determines how often to look for changes in endpoints.            |
```
Nit

crobert-1 commented on extension/observer/cfgardenobserver/config_test.go:
```diff
@@ -26,30 +27,181 @@ func TestLoadConfig(t *testing.T) {
+	cases := []struct {
+		reason string
+		cfg    Config
+		msg    string
```
Just a thought: I see some value in checking specific error messages here, but I'm also concerned that this will cause a lot of noise when needing to update error messages for clarity etc. We could simply switch to a boolean that shows whether we expect and error or not. Your call, whatever you think makes more sense 👍

tomasmota replied:
Had the same thought, but with the different conditionals it's pretty easy to get an error by getting into the wrong branch, so I think we can keep the messages for now and remove later if it turns out to be noisy.

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
 	*observer.EndpointsWatcher
+	cancel context.CancelFunc
+	config *Config
+	ctx    context.Context
```
[Contexts shouldn't be stored in `structs`.](https://go.dev/blog/context-and-structs) Is there some other way to handle the necessary functionality?

tomasmota replied:
Good point, I honestly just did the same as the dockerobserver. Will try to find an alternative.

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+	g.containers = infos
+}
+
+func infoToAppID(info garden.ContainerInfo) (string, error) {
```
I don't really think having this in a separate function provides much value. Callers still have to check the returned error, so it just ends up being a function call vs. accessing the map directly.

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+}
+
+func infoToAppID(info garden.ContainerInfo) (string, error) {
+	id, ok := info.Properties["network.app_id"]
```
It maybe helpful to have a central variable for `network.app_id`.

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+func infoToAppID(info garden.ContainerInfo) (string, error) {
+	id, ok := info.Properties["network.app_id"]
+	if !ok {
+		return "", errors.New("could not find app_id")
```
This error message may not be super helpful to users. What's the relevance to the end user that they could understand?

tomasmota replied:
replaced with `return fmt.Errorf("container properties do not have a '%s' field, required to fetch App labels", propertiesAppIDKey)
`

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+			continue
+		}
+
+		if info.State != "active" {
```
Maybe use a central variable for "active". I see the cloud foundry golang package doesn't have one, so we can just declare our own.

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+// containerEndpoints generates a list of observer.Endpoint for a container,
+// this is because a container might have more than one exposed ports
+func (g *cfGardenObserver) containerEndpoints(handle string, info garden.ContainerInfo) []observer.Endpoint {
+	portsProp, ok := info.Properties["network.ports"]
```
Maybe we can use a central variable for "network.ports"

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+	if g.config.IncludeAppLabels {
+		app, err = g.App(info)
+		if err != nil {
+			g.logger.Error("error fetching Application", zap.Error(err))
```
```suggestion
			g.logger.Error("error fetching application", zap.Error(err))
```
Nit

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+
+	endpoints := []observer.Endpoint{}
+	for _, port := range ports {
+		port, err := strconv.ParseUint(port, 10, 16)
```
We shouldn't have `port` declared twice here. Maybe the variable declared in the `for` line could be `portString` or `portUnformatted`. Something else may be better.

crobert-1 replied:
`err` is already declared, so we don't need the `:=`, maybe we declare `port` above and convert this to a plain `=`

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+//
+// We parse only the tags into a map, to be used as labels
+func parseTags(info garden.ContainerInfo) (map[string]string, error) {
+	logConfig, ok := info.Properties["log_config"]
```
Would it make sense to make `log_config` a central variable as well?

Follow up question for this PR as a whole: Does the cf golang package define any of these properties as constants somewhere that we could use, instead of defining them all ourselves?

tomasmota replied:
I have made all of these central variables now. 

The [garden client](https://pkg.go.dev/code.cloudfoundry.org/garden), which is use to fetch this information, does not provide any constants for these. As far as I can tell these are just properties created by [guardian](https://github.com/cloudfoundry/guardian/) when creating the containers, which doesn't seem to have them as constants either.

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+}
+
+// The info.Properties contains a key called "log_config", which
+// has contents that look like the following JSON endoded string:
```
```suggestion
// has contents that look like the following JSON encoded string:
```

crobert-1 commented on extension/observer/cfgardenobserver/extension.go:
```diff
@@ -4,33 +4,337 @@
+		return nil, fmt.Errorf("error unmarshaling logConfig: %w", err)
+	}
+
+	tags, ok := data["tags"].(map[string]any)
```
Same request, make `tags` a central variable.

crobert-1 reviewed (commented):
Mostly just nits.

_I haven't gotten a chance to look through tests yet, I wanted to get a review out sooner rather than later. Hope it's helpful_ 👍

tomasmota committed:
[extensions/observer/cfgardenobserver] Address review tweaks

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Gen otelcontribcol

tomasmota commented:
> Mostly just nits.
> 
> _I haven't gotten a chance to look through tests yet, I wanted to get a review out sooner rather than later. Hope it's helpful_ 👍

I appreciate that, have addressed all the issues except for the context thing, which I'll try to change later.

tomasmota committed:
[extensions/observer/cfgardenobserver] Keep channel instead of context in struct

tomasmota committed:
[extensions/observer/cfgardenobserver] Remove CancelFunc

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go mod

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Make authType private

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

crobert-1 commented on extension/observer/cfgardenobserver/README.md:
```diff
@@ -20,34 +20,54 @@ The `cfgarden_observer` looks at the current host to discover Garden containers.
+| include_app_labels          | bool   | false                                                                  | Determines whether or not app labels get added to container labels |
+| garden.endpoint             | string | /var/vcap/data/garden/garden.sock                                      | Path to garden socket.                                             |
+| cloud_foundry.endpoint      | string | none. required when `include_app_labels` is set to `true` | CloudFoundry API endpoint                                          |
+| cloud_foundry.auth_type     | string | none. required when `include_app_labels` is set to `true` | Authentication type, one of: user_pass, client_credentials, token  |
```
What do you think of moving authentication options under a specific config section for each? Something like:
```
auth_type: user_pass
user_pass:
  username: "a"
  password: "..."
```
I'm wondering if it would help clarify that each of these options are only relevant for a specific authentication type. If a config then contains options irrelevant to the `auth_type` chosen, they'd just be ignored. Do you think that would be a more clear way of handling this?

tomasmota replied:
hmm it's a good point. I tried implementing your suggestion and after seeing it I personally feel like it's more clunky to have them in separate sections, I would rather make it clearer in the readme.

The following, for example, looks a bit odd to me:
```
auth_type: token
token:
  access_token: "a"
  refresh_token: "..."
```

Other namings I tried for the token stuff also seem odd to me, but maybe this is just a personal preference. Ultimately I value your opinion more so if you still think it's best of course I will do it that way!

crobert-1 replied:
That's fair, I can see that being a bit clunky. Would it make sense to put all authentication options under the same header?

```
authentication:
  type: token
  access_token: "a"
  refresh_token: "..."
```
It could be `auth` instead of `authentication`, but do you think that would help? I just want to avoid having such a flat layout when all of these options are related/dependent on each other somehow.

tomasmota replied:
good idea, I've changed it now

crobert-1 commented on extension/observer/cfgardenobserver/config_test.go:
```diff
@@ -26,30 +27,181 @@ func TestLoadConfig(t *testing.T) {
+	}
+}
+
+func loadConf(t testing.TB, path string, id component.ID) *confmap.Conf {
```
`loadConfig` and `loadConf` are a bit unclear on how they differ and their purpose. Can we rename to make their difference more clear? `loadConf` could be something like `loadConfFromFile`, or something else if you have a better idea.

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Rename loadConf to loadRawConf

tomasmota committed:
[extensions/observer/cfgardenobserver] crosslink

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] gotidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

crobert-1 commented on extension/observer/cfgardenobserver/README.md:
```diff
@@ -20,34 +20,54 @@ The `cfgarden_observer` looks at the current host to discover Garden containers.
+| garden.endpoint             | string | /var/vcap/data/garden/garden.sock                                      | Path to garden socket.                                             |
+| cloud_foundry.endpoint      | string | none. required when `include_app_labels` is set to `true` | CloudFoundry API endpoint                                          |
+| cloud_foundry.auth_type     | string | none. required when `include_app_labels` is set to `true` | Authentication type, one of: user_pass, client_credentials, token  |
+| cloud_foundry.username      | string | none.                                                                   | Username (auth_type: user_pass)                                    |
```
```suggestion
| cloud_foundry.username      | string | none                                                                   | Username (auth_type: user_pass)                                    |
```

crobert-1 commented on extension/observer/cfgardenobserver/config.go:
```diff
@@ -4,15 +4,116 @@
 	RefreshInterval time.Duration `mapstructure:"refresh_interval"`
+
+	// The time to wait before resyncing app information on cached containers
+	// using the CloudFoundry API. Example: cache_sync_interval: "20m"
```
```suggestion
	// using the CloudFoundry API.
```
Nit: Config examples are in the README, it's not necessary to include here as well.

crobert-1 reviewed (commented):
Looking pretty good, just another suggestion on configuration layout, and some nits.

tomasmota committed:
[extensions/observer/cfgardenobserver] add auth section under cloud_foundry

tomasmota committed:
[extensions/observer/cfgardenobserver] remove example from config comments

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

crobert-1 commented:
Failing unit test is unrelated to this change, I've filed #35007.

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
gotidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota commented:
hey @crobert-1 there is this single prometheus compliance test that keeps failing for more than a week now and I have no clue why. I have tried checking if it was something with dependencies but no luck. Do you have any idea what it could be?

crobert-1 commented:
> hey @crobert-1 there is this single prometheus compliance test that keeps failing for more than a week now and I have no clue why. I have tried checking if it was something with dependencies but no luck. Do you have any idea what it could be?

Hey @tomasmota, sorry about that. It's frequency of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35119, unrelated to your changes 👍

crobert-1 commented:
_It looks like some imports are the wrong version again, but I'm adding `ready to merge` as it should be good once that's fixed._ 👍

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
Merge branch 'main' into observer/cfgardenobserver-3

tomasmota committed:
[extensions/observer/cfgardenobserver] Go tidy

tomasmota committed:
[extensions/observer/cfgardenobserver] Remove tomasmota as codeowner

andrzej-stencel commented:
Sorry for the delay @tomasmota. Can you fix the conflicts again :sweat:

andrzej-stencel commented:
@m1rp It looks like this still has conflicts, can you take a look?

smcllw committed:
Merge branch 'main' into observer/cfgardenobserver-3

smcllw committed:
[extensions/observer/cfgardenobserver] go tidy

m1rp commented:
i've also done the CLA authorization now

atoulme commented:
easyCLA still fails, please review

MovieStoreGuy commented:
Would it be possible to squash some of these commits? Hopefully it should make any branch updates easier :)

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

m1rp commented:
EasyCLA has worked this time thankfully!

smcllw committed:
squash commit

smcllw committed:
delete obsolete file

smcllw committed:
go tidy

smcllw committed:
fix merge conflicts

smcllw committed:
add changelog entry that was deleted

smcllw committed:
dependency updates

jriguera committed:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3

Commit:
Merge branch 'main' into observer/cfgardenobserver-3
